misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java

```diff
@@ -176,6 +146,69 @@ private void processEvent(EventSet eventSet, Event event) {
+                breakpoint.getCondition().get() : "";
+        String logMessage = breakpoint.getLogMessage().isPresent() && !breakpoint.getLogMessage().get().isBlank() ?
+                breakpoint.getLogMessage().get() : "";
+
+        if (logMessage.isEmpty() && condition.isEmpty()) {
+            notifyStopEvent(event);
+            return;
+        }
+
+        // If there's an non-empty user defined log message and and no breakpoint condition, resumes the remote VM
+        // after showing the log on the debug console.
+        if (!logMessage.isEmpty() && condition.isEmpty()) {
+            context.getAdapter().sendOutput(logMessage, STDOUT);
+            context.getDebuggeeVM().resume();
+            return;
+        }
+
+        // When evaluating breakpoint conditions, we might need to invoke methods in the remote JVM and it can
+        // cause deadlocks if 'invokeMethod' is called from the client's event handler thread. In that case, the
+        // thread will be waiting for the invokeMethod to complete and won't read the EventSet that comes in for
+        // the new event. If this new EventSet is in 'SUSPEND_ALL' mode, then a deadlock will occur because no one
+        // will resume the EventSet. Therefore to avoid this, we are disabling possible event requests before doing
+        // the condition evaluation.
+        context.getEventManager().classPrepareRequests().forEach(EventRequest::disable);
+        context.getEventManager().breakpointRequests().forEach(BreakpointRequest::disable);
+        CompletableFuture<Boolean> resultFuture = evaluateBreakpointCondition(condition, event.thread());
+        try {
+            Boolean result = resultFuture.get(5000, TimeUnit.MILLISECONDS);
+            if (result) {
+                if (!logMessage.isEmpty()) {
+                    context.getAdapter().sendOutput(logMessage, STDOUT);
+                    // As we are disabling all the breakpoint requests before evaluating the user's conditional expression,
+                    // need to re-enable all the breakpoints before continuing the remote VM execution.
+                    restoreBreakpoints(context.getLastInstruction());
+                    context.getDebuggeeVM().resume();
+                } else {
+                    notifyStopEvent(event);
+                }
+            } else {
+                // As we are disabling all the breakpoint requests before evaluating the user's conditional expression,
+                // need to re-enable all the breakpoints before continuing the remote VM execution.
+                restoreBreakpoints(context.getLastInstruction());
+                context.getDebuggeeVM().resume();
+            }
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            context.getAdapter().sendOutput("Warning: Skipping the conditional breakpoint at line: "
```

Comment:
Shouldn't we consider the 'e' exception's details for the output?

Reply:
@prabushi As we show this warning in the debug console during the debug session, I doubt whether we need to expose java level exception details to the user.

Reply:
Maybe we can add them to debug level logs later.